query-algebrizer/src/validate.rs

```diff
@@ -74,6 +75,46 @@ pub fn validate_or_join(or_join: &OrJoin) -> Result<()> {
     }
 }
 
+pub fn validate_not_join(not_join: &NotJoin) -> Result<()> {
+    // Grab our mentioned variables and ensure that the rules are followed.
+    match not_join.unify_vars {
+        UnifyVars::Implicit => {
+            if not_join.clauses.len() < 2 {
+                Ok(())
+            } else {
+                let mut clauses = not_join.clauses.iter();
+                let template = clauses.next().unwrap().collect_mentioned_variables();
+                for clause in clauses {
+                    if template != clause.collect_mentioned_variables() {
+                        bail!(ErrorKind::NonMatchingVariablesInNotClause);
+                    }
+                }
+                Ok(())
+            }
+        },
+        UnifyVars::Explicit(ref vars) => {
+            // The joined vars must each appear somewhere in the clauses mentioned variables.
+            let var_set: BTreeSet<Variable> = vars.iter().cloned().collect();
+            for var in &var_set {
+                if !&not_join.clauses.iter().any(|clause| clause.collect_mentioned_variables().contains(&var))  {
```

Comment:
As in the `Implicit` case: the parts of a `not` must all match together; they're not considered separately.

Here you want to make sure that `vars` is a subset of the union of the mentioned variables in each clause. You can use `accumulate_mentioned_variables` to accumulate from each, or you can try to figure out a cheaper way to do it.

In any case, you should probably implement `ContainsVariables` for `NotJoin`, then you can just say

```
if !var_set.is_subset(self.collect_mentioned_variables()) {
    …
```